### PR TITLE
fix several bugs in the bot 

### PR DESCRIPTION
**Pull request by rizar**
Now I am dealing with `--shift=485 --seed=5`. The agent can't decide which key to go to!

image

### CODE REVIEW

**Comment by rizar** [comment]
Now it fails on `--seed=9 --shift=3639`

**Comment by rizar** [comment]
The new bug is `--seed=9 --shift=1224` and it is super fun!

The current logic of path-finding is as follows. First, the bot tries to find paths which are blocked. If they don't exist, the bot tries to find a path with blockers, and out of such, it prefers the one in which the first blocker is as a close (in Manhattan metric) to the target as possible. Often, this results in a reasonable behavior, but sometimes it doesn't:

image

Here, the bot has to go to the red door that connects the bottom-right room with the rest of the grid. It doesn't know that there is a path through the purple door because it hasn't been there, so it plans to take a longer route through the top-most rooms. It decides that the blockers that are at least 10 steps (in Manhattan distance) away from the red door are unavoidable, i.e. that the red ball is an unavoidable blocker, which is obviously not true. That is not that dramatic on its own, but the current unblocking heuristic ends up being unhelpful in this case, and the bot enters an endless loop. 

I am not sure 100% how to proceed, because the endless loop here is a result of both suboptimal path finding and suboptimal dropping strategy.

**Comment by rizar** [comment]
I guess the dropping heuristics can never be perfect, but as for the shortest path search, we could implement a proper Dijkstra algorithm to minimize the total number of blockers in the paper. Thoughts?

**Comment by maximecb** [comment]
If the problem is an endless loop, one simple way to fix it would be to move the blockers to a pseudo-random location. Preferably this should be done using an RNG that we seed ourselves so that the behavior is repeatable. We could have an RNG in the bot that we always seed at 0. Otherwise, maybe we should favor exploration before unblocking if the bot hasn't seen much of the current room?

How would you fix this with a proper Dijkstra's implementation?

**Comment by rizar** [comment]
> Otherwise, maybe we should favor exploration before unblocking if the bot hasn't seen much of the current room? 

In this case, the bot has seen enough of the current room, but it hasn't seen others. 

> How would you fix this with a proper Dijkstra's implementation?

The bot would choose the path which has the least blockers on it. It seems to me that for such paths the blocking heuristic would be less likely to fail, i.e. less likely to lead to an endless.

Randomization is another option, indeed.

**Comment by rizar** [comment]
The video for the issue discussed above: https://www.useloom.com/share/8477912a721e4908adb0deaade6b450a

**Comment by rizar** [comment]
Looks like the bot is finally flawless (already generated 500k demos for BossLevel without failures!)

This PR contains:

- A major overhaul of the path-finding mechanism. No more Manhattan distance for choosing which blockers are unavoidable. Now the bot tries to go as far as possible with unblocking. If unblocking is unavoidable, it find a shortest path from the cells that are reachable without unblocking to the goal. 
- the "bye-bye GoToAdjPos" code from #17 
- environment serialization code in `gui.py` that is really helpful when you need to debug something that happens for only big `--shift`'s. Now you can pickle the mission on which the bot fails and not wait for several thousands of resets
- some cosmetic changes in `levelgen.py` that were required to support mission pickling/unpickling
- semantics of `pos_next_to` in `verifier.py` is changed to make more sense
- a lot of regression tests (which I still have to add to Travis)

@maximecb @saleml it would be good to hear your opinion about this by Tuesday (tomorrow I will work on smth else).

I am doing a bit more testing, will report soon.

**Comment by maximecb** [comment]
A better way to avoid the shift thing you and salem have been doing would be to set a different seed for each mission.

**Comment by rizar** [comment]
> A better way to avoid the shift thing you and Salem have been doing would be to set a different seed for each mission.

Maybe. I don't remember the historical reason why we didn't go for that. I would be reluctant to go ahead and change this everywhere though.

The new bot takes a tiny bit more steps on average (86.06 vs 85.34 on 50 episodes of BossLevel). Not sure if I want to investigate this.

I am also planning several more refactoring PRs with the goal of simplifying the code and making it more maintainable.

**Comment by maximecb** [comment]
Could we change it for the bot when generating demos without changing it elsewhere?

**Comment by rizar** [comment]
Off the top of my head, it depends on how test examples are generated
everywhere in the code. I would need to take a look.

**Comment by saleml** [comment]
TLDR: 1 - Awesome. 2 - The fact that the bot takes more steps on average on 50 seeds is okay.

The new `shortest_path` algo is very well thought ! Kuddos !

> The new bot takes a tiny bit more steps on average (86.06 vs 85.34 on 50 episodes of BossLevel)

Amongst the 50 first seeds, only seed 9 has worse performance in this version of the bot (screenshot: the bot has to go to the green ball) - ExploreSubgoal makes the (old bot) c go to the grey door, and makes the bot in this PR go to the red door, which is the **optimal thing to do a priori** (the red door is closer). **Question**: I think, but I am not sure, that the way you implemented the shortest path algo, looking for **an** unopened door leads to the choice of **the closest** one amongst all unopened doors, even if it's not explicitly mentioned in the code, and this is why the red door was chosen in this scneario. Is my understanding correct ? 
If this is the case, then `find_obj_pos` can be simplified as well. We can discuss more in person.
screen shot 2019-01-21 at 19 46 09

In two other seeds, the new bot performs slightly better, thanks to `find_obj_pos` actually picking the object that matches the description that's closest in terms of path length, even if path contains blockers, IIUC. (That was not the case in old bot ! )

P.S.: this bot performs slightly better than the old one using 200 seeds (86.47 vs 86.61) :)

**Comment by rizar** [comment]
Thanks for taking look, @saleml ! Indeed, in the example above it is not obvious apriori behind which door the ball will be located.

I would love to simplify things ever further, including taking a look at `find_obj_pos`. For now, I think it would be good to merge these changes. I will remove the controversial loading and pickling of the environment, and I would rather take a look at switching from resets to seeds a bit later, if this is fine with you, @maximecb . I will even create an issue for that.

The bot still fails on smth like 3 missions out of 100000, I'm trying to investigate. I would merge this PR before figuring this out though, because it's clearly an improvement.

**Comment by maximecb** [comment]
> I would rather take a look at switching from resets to seeds a bit later, if this is fine with you, @maximecb . I will even create an issue for that.

Sounds good.
